onnx/defs/math/defs.cc

```diff
@@ -621,7 +621,101 @@ ONNX_OPERATOR_SCHEMA(MatMul)
         "Constrain input and output types to float tensors.")
     .SetDoc(R"DOC(
 Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html
-)DOC");
+)DOC")
+    .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+        if (!hasNInputShapes(ctx, 2)) {
+          return;
+        }
+
+        auto shape0 = ctx.getInputType(0)->tensor_type().shape();
+        auto shape1 = ctx.getInputType(1)->tensor_type().shape();
+
+        if (shape0.dim_size() == 0 || shape1.dim_size() == 0) {
+          return;
+        }
+
+        TensorShapeProto paddedShapeL;
+        TensorShapeProto paddedShapeR;
+
+        for (int i = 0; i < shape1.dim_size() - std::max(2, shape0.dim_size()); ++i) {
+          paddedShapeL.add_dim()->set_dim_value(1);
+        }
+        for (int i = 0; i < shape0.dim_size() - std::max(2, shape1.dim_size()); ++i) {
+          paddedShapeR.add_dim()->set_dim_value(1);
+        }
+
+        if (shape0.dim_size() == 1) {
+          paddedShapeL.add_dim()->set_dim_value(1);
+          *paddedShapeL.add_dim() = shape0.dim(0);
+        } else {
+          for (int i = 0; i < shape0.dim_size(); ++i) {
+            *paddedShapeL.add_dim() = shape0.dim(i);
+          }
+        }
+
+        if (shape1.dim_size() == 1) {
+          *paddedShapeR.add_dim() = shape1.dim(0);
+          paddedShapeR.add_dim()->set_dim_value(1);
+        } else {
+          for (int i = 0; i < shape1.dim_size(); ++i) {
+            *paddedShapeR.add_dim() = shape1.dim(i);
+          }
+        }
+
+        auto dimSize = paddedShapeL.dim_size();
+
+        if (paddedShapeR.dim_size() != dimSize) {
+          return;
+        }
+
+        {
+          // check for compatible matrix dimensions
+          auto dimL = paddedShapeL.dim(dimSize - 1);
+          auto dimR = paddedShapeR.dim(dimSize - 2);
+          if (dimL.has_dim_value() && dimR.has_dim_value() &&
+              dimL.dim_value() != dimR.dim_value()) {
+            return;
+          }
+        }
+
+        TensorShapeProto resultShape;
+
+        for (int i = 0; i < dimSize - 2; ++i) {
```

Comment:
Looks good to me ... I guess the broadcasting-logic alone may be reusable across multiple broadcasting ops, if we can factor it out. (I haven't checked the other broadcasting ops yet.)